### PR TITLE
reporting on compiler and runtime stats

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Extended support to make report to work with compile time stats (Simd) or runtime (Performance profiling)

Manual
```
make-report -[svh] [-c <compile_log>] [-r <run_log>] [-l <num>] [-p <op regexp>]

Usage: Report statistics on compiler and runtime characteristics of onnx ops.

Compile time statistics are collected from the `onnx-mlir` compiler output
with the option `--onnx-op-report` equal to `Simd` or other supported sub-options.

Runtime statistics are collected from the runtime output of a model compiled.
with the option `--profile-ir` equal to `Onnx` or other supported sub-options

When both compile time and runtime statistics are provided, the script
will correlate the performance metrics with data gathered at compile time

Additional help:
  If you need more specific info on individual success/failure, run 
  `onnx-mlir --debug-only=lowering-to-krnl` and look at the compiler output.
  Use `-d 3` to correlate the node name printed here with compiler output.

Parameters:
  -c/--compile <file_log>: File name containing the compile time statistics.
  -r/--runtime <file_log>: File name containing the runtime statistics.
  -l/--level <num>:        Print statistics:
                           0: Just count successful/unsuccessful ops.
                           1: Also count reasons for success/failure.
                           2: Also list metrics.
                           3: Also list node name.
  -f/--focus <regexp>:     Focus only on ops that match the regexp pattern.
  -s/supported:            Focus only on ops that are supported, i.e. skip unsupported ops.
  -v/--verbose:            Run in verbose mode.
  -h/--help:               Print usage.
```

Example where the stats from SIMD are related to performance run, here for bertsquad10 on Mac M1
```
% make-report.py  -r new-times2.log -c new-compile2.log -l 1 -s
Statistic legend:
  op-name: num, time(s)

Statistics (ignore unsupported ops):
   onnx.Add-simd : 112, 0.0133830
     *, 0.0133830 : variadic with manageable broadcast
   onnx.Cast : 23, 0.0002040
     21, 0.0001910 : no simd in unary because could not find beneficial VL
     2, 0.0000130 : no simd in unary because scalar/layouts
   onnx.Gemm : 1, 0.0003550
     *, 0.0003550 : no simd because of small j trip count
   onnx.Gemm-simd : 72, 0.7421320
     *, 0.7421320 : simd directly in output tiles
   onnx.MatMul-simd : 25, 0.0631070
     *, 0.0631070 : simd for mat * mat along j dim
   onnx.Mul : 36, 0.0004160
     *, 0.0004160 : no simd in variadic because no manageable broadcast/layout 
   onnx.Mul-simd : 198, 0.0267680
     *, 0.0267680 : variadic with manageable broadcast
   onnx.Reciprocal-simd : 25, 0.0001400
     *, 0.0001400 : unary fully flattened
   onnx.ReduceMeanV13-simd : 50, 0.0008720
     *, 0.0008720 : horizontal
   onnx.Sqrt-simd : 25, 0.0001310
     *, 0.0001310 : unary fully flattened
   onnx.Sub-simd : 62, 0.0018620
     *, 0.0018620 : variadic with manageable broadcast
   onnx.Tanh : 12, 0.0052510
     *, 0.0052510 : no simd in unary because could not find beneficial VL
```

More details can be gathered with higher detail levels, up to listing the node names of each op for correlating with specific instance of the op.

Running the script on the performance only:
```
% make-report.py  -r new-times2.log                            
Statistic legend:
  op-name: num, cumulative time(s)

Statistics start (all ops):
  onnx.Add, 112, 0.0133830
  onnx.Cast, 105, 0.0009170
  onnx.Concat, 55, 0.0004700
  onnx.Constant, 473, 0.0099610
  onnx.ConstantOfShape, 1, 0.0000110
  onnx.Dim, 5, 0.0000450
  onnx.Gather, 1, 0.0002890
  onnx.Gemm, 73, 0.7424870
  onnx.MatMul, 25, 0.0631070
  onnx.Mul, 235, 0.0427710
  onnx.NoValue, 1, 0.0000130
  onnx.OneHot, 1, 0.0000110
  onnx.Reciprocal, 25, 0.0001400
  onnx.ReduceMeanV13, 50, 0.0008720
  onnx.Reshape, 70, 0.0004920
  onnx.Softmax, 12, 0.0387020
  onnx.Split, 1, 0.0000060
  onnx.Sqrt, 25, 0.0001310
  onnx.Squeeze, 4, 0.0000250
  onnx.Sub, 62, 0.0018620
  onnx.Tanh, 12, 0.0052510
  onnx.Transpose, 49, 0.0050080
  onnx.Unsqueeze, 13, 0.0001110
Statistics end
```

There were some strange behavior in the printing of the node name. In KrnlInstrument.cpp, we create node names in full (using several alternative when no data is available). The symbols are added fine, but it appears that when the code run, the printing of the node names often append the previous node name. Not all the time, but often enough. At other time, it append the op name. The `make-report.py` catches these strange cases and print a warning indicating that they were caught. 

An additional warning is printed if there are ops (from the compile time statistics) for which we don't find their computation time.

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
Implementing issue #2447

File: utils/make-report.py [hunk 1]
```diff
@@ -16,170 +16,387 @@
...
+    print("Parameters:")
+    print("  -c/--compile <file_log>: File name containing the compile time statistics.")
+    print("  -r/--runtime <file_log>: File name containing the runtime statistics.")
+    print("  -l/--level <num>:        Print statistics:")
+    print("                           0: Just count successful/unsuccessful ops.")
+    print("                           1: Also count reasons for success/failure.")
+    print("                           2: Also list metrics.")
+    print("                           3: Also list node name.")
+    print("  -f/--focus <regexp>:     Focus only on ops that match the regexp pattern.")
+    print("  -s/supported:            Focus only on ops that are supported, i.e. skip unsupported ops.")
```
Comment (tungld) [inline utils/make-report.py:53]: Could you explain more about supported/unsupported ops to make them clear? I guess they are related to SIMD and parallel.

**Comment by tungld** [comment]
> The symbols are added fine, but it appears that when the code run, the printing of the node names often append the previous node name.

Not 100% sure if it is related to #2448, but I saw the same behavior with models containing onnx.NoValue.

File: utils/make-report.py (hunk 1 above)
Comment (AlexandreEichenberger) [reply utils/make-report.py:53]: great suggestion
```
  -s/supported:        Focus only on ops that are supported. Namely, the report
                       will skip ops for which compile-time statistics list
                       the 'unsupported' keyword in its printout.
                       For SIMD/parallel statistics, this include all ops that
                       have currently no support for it.
```

**Review by tungld — APPROVED** [review]
LGTM! Thanks for another helpful feature!

**Pull request merged**